core/restorer.go

```diff
@@ -229,18 +250,58 @@ func (r *Restorer) CheckRestorable() (*PrecheckResult, error) {
 	}
 
 	return &PrecheckResult{
-		BackupDate:          backup.BackupCreated,
-		ControllerModelUUID: backup.ControllerModelUUID,
-		JujuVersion:         backup.JujuVersion,
-		ModelCount:          backup.ModelCount,
+		BackupDate:            backup.BackupCreated,
+		ControllerModelUUID:   backup.ControllerModelUUID,
+		BackupJujuVersion:     backup.JujuVersion,
+		ControllerJujuVersion: controller.JujuVersion,
+		ModelCount:            backup.ModelCount,
 	}, nil
 }
 
 // Restore replaces the database's contents with the data from the
 // backup's database dump.
 func (r *Restorer) Restore(logPath string, includeStatusHistory bool) error {
-	err := r.db.RestoreFromDump(r.backup.DumpDirectory(), logPath, includeStatusHistory)
-	return errors.Trace(err)
+	controller, err := r.db.ControllerInfo()
+	if err != nil {
+		return errors.Annotate(err, "getting controller info")
+	}
+	metadata, err := r.backup.Metadata()
+	if err != nil {
+		return errors.Annotatef(err, "getting backup metadata")
+	}
+	logger.Debugf("restoring dump")
+	err = r.db.RestoreFromDump(r.backup.DumpDirectory(), logPath, includeStatusHistory)
+	if err != nil {
+		return errors.Annotatef(err, "restoring dump from %q", r.backup.DumpDirectory())
+	}
+	if controller.JujuVersion != metadata.JujuVersion {
+		logger.Debugf("updating controller agent versions to %s", metadata.JujuVersion)
+		results := r.manageAgents(true, true, func(n ControllerNode) error {
+			logger.Debugf("    %s", n)
+			err := n.UpdateAgentVersion(metadata.JujuVersion)
+			return errors.Annotatef(err, "updating %s", n)
+		})
+		if err := collectMachineErrors(results); err != nil {
+			return errors.Annotatef(err, "problems updating controllers to version %q", metadata.JujuVersion)
+		}
+	}
+	return nil
```

Comment:
How do we roll back if there's errors?
Should we do a prepare check to ensure all controller machines are ssh accessible first before restoring the db dump at least?

Reply:
Yup - we do this, see `cmd/restore.go:220`. 
If it fails then the user will need to manually change the symlink and config file on that machine and start the machine agent. I could add a message pointing to a section in the readme?

Reply:
I'll do this in a followup.